### PR TITLE
[econstr] Flag to make `to_constr` fail if its output contains evars

### PR DESCRIPTION
**Pull request by ejgallego**
We forbid calling `EConstr.to_constr` on terms that are not evar-free,
as to progress towards enforcing the invariant that `Constr.t` is
evar-free. [c.f. #6308]

Due to compatibility constraints we provide an optional parameter to
`to_constr`, `abort` which can be used to overcome this restriction
until we fix all parts of the code.

Now, grepping for `~abort:false` should return the questionable
parts of the system.

Not a lot of places had to be fixed, some comments:

- problems with the interface due to `Evd/Constr` [`Evd.define` being
  the prime example] do seem real!
- inductives also look bad with regards to `Constr/EConstr`.
- code in plugins needs work.

A notable user of this "feature" is `Obligations/Program` that seem to
like to generate kernel-level entries with free evars, then to scan
them and workaround this problem by generating constants.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Why is the commit from #6422 appearing?

**Comment by SkySkimmer** [comment]
I think `abort` should be called `allow_undefined_evars` or some such name which mentions evars.

File: vernac/himsg.ml
Comment (SkySkimmer) [inline vernac/himsg.ml]: Why not use pr_leconstr here?

Comment (ejgallego) [reply vernac/himsg.ml]: Indeed, I was hesitating at first whether to make this PR "tagging-only" or to also fix some trivial ocurrences. I guess it makes sense to fix a few, so I'll update this + I will change the name of the flag.

**Comment by ejgallego** [comment]
> I think abort should be called allow_undefined_evars or some such name which mentions evars.

Good point, I wanted a very scary name as the idea is for the flag to go away, I'll give it another try.

**Comment by maximedenes** [comment]
It seems the linter fails.

**Comment by ejgallego** [comment]
Note that this PR implements a check that should be enable in dev builds, but should be disabled in production IMO. So I dunno how to handle such stuff.

**Comment by SkySkimmer** [comment]
>Note that this PR implements a check that should be enabled in dev builds, but should be disabled in production IMO. So I dunno how to handle such stuff.

Resurrect configure time -debug?
Instead of `if abort then anomaly else constr.map ` use `assert (not abort); constr.map` and let production compilers use the option for removing asserts?

**Comment by maximedenes** [comment]
@ejgallego What is the status of this w.r.t. @SkySkimmer 's comment?

**Comment by ejgallego** [comment]
> @ejgallego What is the status of this w.r.t. @SkySkimmer 's comment?

I am not sure what would be the best option about how to handle this flag. `abort` is less informative; checking `-debug` would be not good as nobody would test this then. I dunno...

This is again a kind of setting we want to enable for development and on all CI, but that could be annoying to have in production, will think.

A possibility is merging as-is, then removing the fatality of the check as soon as you branch 8.8.

**Comment by maximedenes** [comment]
> A possibility is merging as-is, then removing the fatality of the check as soon as you branch 8.8.

I'd rather not add tasks on the 8.8 todo-list, I'm scared of not being able to complete all of them.

**Comment by ejgallego** [comment]
Well in that case we'd open a blocker bug as a reminder. We are talking about a one line diff thou. Maybe indeed Enrico's latest work on profiles can help with this of stuff. But again this PR is a bit similar to #6343 , in the sense that it tries to check if some parts of the code are doing strange stuff.

**Comment by ejgallego** [comment]
We are too close to the freeze, and this PR is indeed the kind of thing you want to use in a dev cycle, so instead I suggest we merge this [with the strong failure semantics] at the beginning of the 8.9 cycle.

**Comment by maximedenes** [comment]
I think the maintainer mainly concerned by this change is @ppedrot, so putting him as an assignee. @ppedrot, please read `dev/doc/MERGING.md` to know more about the new merging process.

**Comment by ejgallego** [comment]
Indeed Pierre-Marie is the principal person here.

File: pretyping/cases.ml
Comment (gares) [inline pretyping/cases.ml]: no (in theory), only CI can tell in practice

**Review by gares — APPROVED** [review]

**Comment by ejgallego** [comment]
There are a couple of cases missing due to the merge of some PRs not being accounted by this one; will look into it later on.

After this PR is in, we can pinpoin places needing cleanup by their use of `~abort_on_undefined_evars`, I didn't want to do the cleanup in this PR itself as to keep the separation of concerns.

**Comment by ppedrot** [comment]
We cannot merge this PR as-is while the relevant failures in the CI have not been fixed. What is the battle plan?

**Comment by ejgallego** [comment]
> We cannot merge this PR as-is while the relevant failures in the CI have not been fixed. What is the battle plan?

As I said in the comment above the failures are due to some merges / overlays not being updated yet, I'll do a pass soon.

The PR has already passed CI a few times times in the past so I don't expect a significant problem here.

**Review by aspiwack — APPROVED** [review]
Looks all reasonable to me.

File: vernac/comFixpoint.ml
Comment (mattam82) [inline vernac/comFixpoint.ml]: Yes, I guess we should rather have an interface change in obligations.mli so that it takes EConstr.t's there.

File: vernac/comProgramFixpoint.ml
Comment (mattam82) [inline vernac/comProgramFixpoint.ml]: Same here. Obligations turns the undefined evars into obligations, it should rather take EConstr.t's

**Comment by herbelin** [comment]
What procedure to resign as a reviewer? Enough people are on this PR and already made a review. I don't see what I can bring more (or, if I had to say something, I would only suggest that the new flag name is maybe slightly shorter, e.g. abbreviating "undefined" into "undef", but that's not so clear either as full names are easier to remember than abbreviated names; on the other side of the other side, auto-completion is able to remember, so abbreviating is also ok).

So, is there a way to resign, giving a motivation, but not cluttering mailboxes? I guess no.

**Comment by ejgallego** [comment]
> What procedure to resign as a reviewer? Enough people are on this PR and already made a review. I don't see what I can bring more (or, if I had to say something, I would only suggest that the new flag name is maybe slightly shorter, e.g. abbreviating "undefined" into "undef", but that's not so clear either as full names are easier to remember than abbreviated names; on the other side of the other side, auto-completion is able to remember, so abbreviating is also ok).

Well, in this case this is one of the few PRs that actually every component owner should be aware. Almost every use of the flag indicates a strange situation in the subsystem so in a sense flag use == todo for the subsystem.

**Review by herbelin — COMMENTED** [review]
I shall try to propose a flag-free version for cctac and firstorder. This led me to read parts of the PR more carefully and to give some comments.

File: engine/eConstr.mli
Comment (ejgallego) [reply engine/eConstr.mli]: thanks for the comment @herbelin , will be taken into account in the next revision.

**Comment by ejgallego** [comment]
> Should this be benched?

The commit was at some point [with no discernible effect], but sure, let's bench again.

**Comment by herbelin** [comment]
About @ejgallego's questions:

> - inductives also look bad with regards to Constr/EConstr.

@Thierry-Martinez precisely wondered yesterday why `inductive_family` was not using `EConstr.constr` and he might change it. Did you have that in mind?

> - code in plugins needs work.

For `cc`, a hash table is used to identify common subterms. These subterms possibly contain evars but the hashing API is on `constr`. I see at least the following options (signalling also @ppedrot about the general architecture, and @PierreCorbineau who accepted being maintainer of `cctac`):
1. considering that it is legitimate to have evars in `Constr.t` to consider that `to_constr ~abort_on_undefined_evars:false` as legitimate for some cases
2. idem but adding an explicit `to_open_constr` to do what `to_constr ~abort_on_undefined_evars:false` does
3. to add a new type, say `Constr.repr` to intend the low-level representation of `Constr.t` in which evars are allowed, and have hashing functions working on it
4. to add a hashing API in `EConstr` hashing pairs `evar_map * EConstr.t` with, say, equality `eq (sigma1,c1) (sigma2,c2) = assert (sigma1 == sigma2); EConstr.eq_constr_nounivs sigma1 c1 c2`

For `firstorder`, there is a similar issue but with `Set` of `Constr.t` possibly containing evars.

I'm leaning towards experimenting 4. Any opinions?

**Comment by mattam82** [comment]
> @thierry-martinez precisely wondered yesterday why inductive_family was not using EConstr.constr and he might change it. Did you have that in mind?

Definitely a more complete move of inductiveops to EConstr.t is in order.

> cc

If the tactic does no unification of evars and only of universes (which I think is the case AFAIR) then it should be globally parameterized by an evar_map and handle EConstr.t's I think. Can the design be adapted?

**Comment by herbelin** [comment]
> If the tactic does no unification of evars and only of universes (which I think is the case AFAIR)

Right, probably.

> should be globally parameterized by an evar_map and handle EConstr.t's I think.

I don't see what you mean. Do you mean to declare a local hashtable module within the scope of a function depending on sigma?

**Comment by ppedrot** [comment]
Who's supposed to merge this PR? Me, as indicated by the assignation, or @gares, as indicated by the project tagging? The first one that does it?

**Comment by ppedrot** [comment]
Sorry, I probably read too quickly "Enrico's PRs"! Anyways, this is still waiting for benching, I'll launch one.

**Comment by ejgallego** [comment]
Your job failed, relaunched: https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/379/console

**Comment by ejgallego** [comment]
As the previous bench showed this seems to have no effect:
```
┌──────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                          │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                          │                         │                                       │                                       │                         │                 │
│             package_name │     NEW     OLD PDIFF   │            NEW            OLD PDIFF   │            NEW            OLD PDIFF   │     NEW     OLD PDIFF   │ NEW OLD PDIFF   │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│         coq-fiat-parsers │  702.64  705.56 -0.41 % │  1944242465089  1952022781881 -0.40 % │  2968441445172  2974898695031 -0.22 % │ 3375280 3368228 +0.21 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-ssreflect │   44.52   44.70 -0.40 % │   122090496522   122127230081 -0.03 % │   142605909812   142738198718 -0.09 % │  537592  537720 -0.02 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-solvable │  202.61  203.34 -0.36 % │   560131475175   562892689583 -0.49 % │   766851472264   767845233678 -0.13 % │  867640  866844 +0.09 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│     coq-mathcomp-algebra │  185.63  186.21 -0.31 % │   513222460316   514758680688 -0.30 % │   689607598325   690090649915 -0.07 % │  644852  652380 -1.15 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-bignums │   79.99   80.23 -0.30 % │   220843409433   221385111565 -0.24 % │   284845475563   285943160726 -0.38 % │  525836  527212 -0.26 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-field │  464.46  465.51 -0.23 % │  1289456791636  1292238707089 -0.22 % │  2060663067397  2061431082946 -0.04 % │  814576  814472 +0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-character │  275.74  276.23 -0.18 % │   764401132191   765308949205 -0.12 % │  1081752567818  1081720209372 +0.00 % │ 1072892 1072804 +0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-fingroup │   64.42   64.52 -0.15 % │   177510316554   177269521890 +0.14 % │   235816900704   236075600008 -0.11 % │  593696  591044 +0.45 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                  coq-vst │ 3988.55 3990.72 -0.05 % │ 11087137041733 11095966829982 -0.08 % │ 14507392343260 14507541264340 -0.00 % │ 2227452 2224492 +0.13 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-flocq │   63.16   63.18 -0.03 % │   174398751914   174213099526 +0.11 % │   211511001686   211764370148 -0.12 % │  661832  663840 -0.30 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-geocoq │ 3111.72 3109.22 +0.08 % │  8633385894577  8623069232610 +0.12 % │ 13803039094334 13801050240388 +0.01 % │ 1213332 1241464 -2.27 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-odd_order │ 1443.75 1442.21 +0.11 % │  4009861380670  4005594220364 +0.11 % │  6658548254231  6665233247749 -0.10 % │ 1360628 1411664 -3.62 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-formal-topology │   37.50   37.41 +0.24 % │   100469445456   100500075270 -0.03 % │   125471108773   125555941678 -0.07 % │  483668  483556 +0.02 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│             coq-compcert │  869.50  866.16 +0.39 % │  2408779668240  2400423285065 +0.35 % │  3429745544654  3429428213504 +0.01 % │ 1330484 1322400 +0.61 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-fiat-core │  111.46  110.86 +0.54 % │   309292551792   308416850840 +0.28 % │   377115991464   376846739602 +0.07 % │  502324  502368 -0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│ coq-mathcomp-real_closed │  172.54  171.26 +0.75 % │   477293263956   474341816485 +0.62 % │   700157449499   700537525363 -0.05 % │  838244  843072 -0.57 % │   0   0  +nan % │
└──────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘

```

**Review by ppedrot — APPROVED** [review]
OK for me.

**Comment by ppedrot** [comment]
@mattam82 Do you agree with the PR as it is currently? The changes in Obligations ought to be deferred to another PR, right?

@PierreCorbineau This PR is marginally related to Funind, do you want to make a particular comment on it anyways?

**Review by maximedenes — APPROVED** [review]

**Review by PierreCorbineau — APPROVED** [review]
I'm ok with it.

**Comment by ppedrot** [comment]
@mattam82 ping!

**Comment by mattam82** [comment]
Yes, indeed, the changes should be deferred to antother PR. Green light on!

**Comment by ejgallego** [comment]
Cool! The effects of this patch are likely to be observed in the wild, so keep an eye folks.

**Comment by SkySkimmer** [comment]
Shouldn't this have had a dev/changes entry? It seems like something plugin developers should know about, and ocaml types won't tell them about it.

**Comment by ppedrot** [comment]
@SkySkimmer Not sure if it really deserves it, we keep changing the semantics of the ML API without care and so far people haven't punched us in the face yet. If you're concerned you can open another PR though...

**Pull request merged**
